### DISCUSSION TITLE
[AJAX] Invalid JSON-RPC

### DISCUSSION THREAD

**Discussion by garrylie** [jcubic/jquery.terminal#957] [General]
I got this error each time I send a command:
```
[AJAX] Invalid JSON-RPC - Server responded:
{"id":"1","result":"pong","jsonrpc":"2.0"}
```
What is wrong with my response?
content-type is application/json.

**Comment by jcubic** [comment]
Do you have a working demo, or can you at least the code you're using?

**Reply by garrylie** [reply]
html:
```
<!DOCTYPE html>
<html>
<head>
    <script src="http://localhost:8000/js/jquery-3.7.1.min.js"></script>
    <script src="http://localhost:8000/js/jquery.terminal.min.js"></script>
    <link rel="stylesheet" href="http://localhost:8000/css/jquery.terminal.min.css"/>
</head>
<body>
    <script src="http://localhost:8000/js/terminal/init.js"></script>
</body>
</html>

```
init.js:
```
$('body').terminal("/rpc", {
    greetings: 'Welcome!'
});
```
/rpc returns application/json:
`{"id":"2","result":"pong","jsonrpc":"2.0"}`

**Reply by jcubic** [reply]
And how your server look like?

**Reply by jcubic** [reply]
if `/rpc` just return `{"id":"2","result":"pong","jsonrpc":"2.0"}` then the error come from `system.describe` that is executed to get list of methods. You can disable the discovery with option: `ignoreSystemDescribe: true`.

`system.describe` was a feature of JSON-RCP 1.1.

**Reply by garrylie** [reply]
I've already tried with "ignoreSystemDescribe: true" and without it. Same result.
Server is just new php laravel project with one of json-rpc lib from your list. I've tried several. All got the same result.
Server just returns 'pong';

**Reply by jcubic** [reply]
Can you create a screenshot? I think that without having the project live, nothing can do.

**Reply by jcubic** [reply]
The same code is working on the website:

https://terminal.jcubic.pl/rpc-demo.html

The only difference is that in your response, there are no `"error"` field.

**Comment by jcubic** [comment]
So the error is that id in your response is a string not a number. This is a function that checks if the response is valid JSON-RPC:

```javascript
function validJSONRPC(response) {
    return typeof response.id === 'number' &&
        (typeof response.result !== 'undefined' ||
         typeof response.error !== 'undefined');
}
```

I will create a bug for this.

**Reply by jcubic** [reply]
- https://github.com/jcubic/jquery.terminal/issues/958

**Reply by garrylie** [reply]
finally man!! Thanks! It works. Turning ID into Integer worked! Happy holidays!

**Reply by jcubic** [reply]
In [JSON-RPC spec](https://www.jsonrpc.org/specification) the id is a number. So this is not valid JSON-RPC. You can report this to the project, but I will make the type checking more relaxed so it will work with strings that are numeric.

**Reply by jcubic** [reply]
Ok I will close the issue then, there's no point of having invalid JSON-RPC valid.
